### PR TITLE
Use faster set for column IDs in schemaexpr

### PR DESCRIPTION
**Pull request by mgartner**
#### sqlbase: add ColSet wrapper for util.FastIntSet of ColumnID

There are numerous places where a `map[sqlbase.ColumnID]struct{}` or a
`util.FastIntSet` is used to represent a set of `sqlbase.ColumnID`. This
commit adds a typed wrapper around `util.FastIntSet` which is an
efficient and ergonimic replacement for maps and `util.FastIntSet`.

Release note: None

#### schemaexpr: use sqlbase.ColSet instead of maps

This commit replaces maps used as sets of integers with sqlbase.ColSet
because it is a more efficient set implementation.

Release note: None

### CODE REVIEW

**Review by RaduBerinde — APPROVED** [review]
:lgtm:

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/49819#-:-M8sPZB9Chp0T5TjTNc-:bnfp4nl)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @mgartner, @mjibson, @RaduBerinde, and @rytaft)

___
*[pkg/sql/schemaexpr/column.go, line 154 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/49819#-M8sOLKH0vPeQQjdoH2e:-M8sOLKH0vPeQQjdoH2f:bislwmq) ([raw file](https://github.com/cockroachdb/cockroach/blob/ce4e23c22d0b33adaffa042a7a1a3bcfe284adbb/pkg/sql/schemaexpr/column.go#L154)):*
> ```Go
> func replaceVars(
> 	desc *sqlbase.MutableTableDescriptor, rootExpr tree.Expr,
> ) (tree.Expr, util.FastIntSet, error) {
> ```

I'd name this return arg so it's more evident what it is, e.g. `_ tree.Expr, columnIDs util.FastIntSet, _ error`
___
*[pkg/sql/schemaexpr/computed_column.go, line 61 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/49819#-M8sOtpY5xbzIHZhQFwF:-M8sOtpY5xbzIHZhQFwG:b-2zgskv) ([raw file](https://github.com/cockroachdb/cockroach/blob/ce4e23c22d0b33adaffa042a7a1a3bcfe284adbb/pkg/sql/schemaexpr/computed_column.go#L61)):*
> ```Go
> 	}
> 
> 	var dependencies util.FastIntSet
> ```

[nit] Unfortunately it's not easy to tell what the numbers in the set are, maybe `depColumnIDs`. Or, if there is or we expect wider-spread usage of sets of sqlbase.ColumnID, we may want to make a nicely typed wrapper like `opt.ColSet`.

**Review by mgartner — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/49819)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @mjibson, @RaduBerinde, and @rytaft)

___
*[pkg/sql/schemaexpr/column.go, line 154 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/49819#-M8sOLKH0vPeQQjdoH2e:-M8uebiF1qvb7LXafX8n:bh70tl) ([raw file](https://github.com/cockroachdb/cockroach/blob/ce4e23c22d0b33adaffa042a7a1a3bcfe284adbb/pkg/sql/schemaexpr/column.go#L154)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

I'd name this return arg so it's more evident what it is, e.g. `_ tree.Expr, columnIDs util.FastIntSet, _ error`
</blockquote></details>

Now that I've added a typed wrapper, I'm thinking this is OK to be unnamed. Do you agree?
___
*[pkg/sql/schemaexpr/computed_column.go, line 61 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/49819#-M8sOtpY5xbzIHZhQFwF:-M8ucVlA27ZLrbhzwT6z:bldmcre) ([raw file](https://github.com/cockroachdb/cockroach/blob/ce4e23c22d0b33adaffa042a7a1a3bcfe284adbb/pkg/sql/schemaexpr/computed_column.go#L61)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

[nit] Unfortunately it's not easy to tell what the numbers in the set are, maybe `depColumnIDs`. Or, if there is or we expect wider-spread usage of sets of sqlbase.ColumnID, we may want to make a nicely typed wrapper like `opt.ColSet`.
</blockquote></details>

Good point. I've updated the name, and I've also created the `sqlbase.ColSet` typed wrapper, since it looks like there's quite a few places that could use it.

Is that name good, or would `sqlbase.ColIdSet` or `sqlbase.ColumnIDSet` be better?

My quick searching of some places this could be used:

**Comment by RaduBerinde** [comment]
___
*[pkg/sql/schemaexpr/computed_column.go, line 61 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/49819#-M8sOtpY5xbzIHZhQFwF:-M8utMUg0JjkOPv2uXKD:bahfr8d) ([raw file](https://github.com/cockroachdb/cockroach/blob/ce4e23c22d0b33adaffa042a7a1a3bcfe284adbb/pkg/sql/schemaexpr/computed_column.go#L61)):*
<details><summary><i>Previously, mgartner (Marcus Gartner) wrote…</i></summary><blockquote>

Good point. I've updated the name, and I've also created the `sqlbase.ColSet` typed wrapper, since it looks like there's quite a few places that could use it.

Is that name good, or would `sqlbase.ColIdSet` or `sqlbase.ColumnIDSet` be better?

My quick searching of some places this could be used:





</blockquote></details>

Nice! Maybe TableColSet? because you can only reasonably use it for columns  of a single table (ids are unique only within each table)

File: pkg/sql/schemaexpr/column.go [hunk 1]
```diff
@@ -163,3 +123,128 @@ func FormatColumnForDisplay(
...
+	return err
+}
+
+// dummyColumn represents a variable column that can type-checked. It is used
+// in validating check constraint and partial index predicate expressions. This
+// validation requires that the expression can be both both typed-checked and
+// examined for variable expressions.
+type dummyColumn struct {
+	typ *types.T
+	// name is only used for error-reporting.
```
Comment (rohany) [inline pkg/sql/schemaexpr/column.go:173]: This comment is out of date now -- see the comments from where it was moved from in `sqlbase`

File: pkg/sql/sqlbase/table_col_set.go [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0, included in the file
+// licenses/APL.txt.
+
+package sqlbase
+
+import "github.com/cockroachdb/cockroach/pkg/util"
+
+// TableColSet efficiently stores an unordered set of column ids.
+type TableColSet struct {
```
Comment (rohany) [inline pkg/sql/sqlbase/table_col_set.go:16]: What is the reason for wrapping util.FastIntSet here? As far as I can tell all the methods just call into `util.FastIntSet`.

Comment (RaduBerinde) [reply pkg/sql/sqlbase/table_col_set.go:16]: It's so it's strongly typed, which makes it evident what it stores and you don't have to cast ColumnIDs to ints and back.

Comment (rohany) [reply pkg/sql/sqlbase/table_col_set.go:16]: I guess, but it also feels like a good amount of duplication.

Comment (RaduBerinde) [reply pkg/sql/sqlbase/table_col_set.go:16]: After working without a wrapper for opt.ColSet for a few releases (until Go implemented mid-stack inlining) I can tell you that the duplication is worth it.

File: pkg/sql/schemaexpr/column.go (hunk 1 above)
Comment (mgartner) [reply pkg/sql/schemaexpr/column.go:173]: Done.

File: pkg/sql/sqlbase/table_col_set.go (hunk 2 above)
Comment (rohany) [reply pkg/sql/sqlbase/table_col_set.go:16]: Sounds good -- i haven't had to use this extensively to really have an opinion.

Comment (mgartner) [reply pkg/sql/sqlbase/table_col_set.go:16]: I think the bit of duplication is worth it for the clarity and conciseness it provides when using `TableColSet`.

**Review by rytaft — APPROVED** [review]
:lgtm: 

> Reviewed 5 of 5 files at r2, 5 of 6 files at r3, 2 of 2 files at r4.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/49819#-:-M8wKaO89JADfmEemud0:bg7igkv)* status: :shipit: complete! 1 of 0 LGTMs obtained (and 1 stale) (waiting on @mgartner, @mjibson, @RaduBerinde, and @rohany)

**Pull request merged**